pkg/build/registry/rest.go

```diff
@@ -0,0 +1,52 @@
+package registry
+
+import (
+	"fmt"
+	"time"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+
+	"github.com/openshift/origin/pkg/build/api"
+)
+
+// ErrUnknownBuildPhase is returned for WaitForRunningBuild if an unknown phase is returned.
+var ErrUnknownBuildPhase = fmt.Errorf("unknown build phase")
+
+// WaitForRunningBuild waits until the specified build is no longer New or Pending. Returns true if
+// the build became ter within timeout, false if it did not, and an error if any other error state occurred.
+// The last observed Build state is returned.
+func WaitForRunningBuild(watcher rest.Watcher, ctx kapi.Context, build *api.Build, timeout time.Duration) (*api.Build, bool, error) {
+	fieldSelector := fields.Set{"metadata.name": build.Name}.AsSelector()
+	w, err := watcher.Watch(ctx, labels.Everything(), fieldSelector, build.ResourceVersion)
+	if err != nil {
+		return nil, false, err
```

Comment:
The intent of this method is to retrieve the latest build status.  A caller
*may* need other parts of the build, and shouldn't have to fetch the build
again.

On Thu, Oct 1, 2015 at 12:27 PM, Michail Kargakis <notifications@github.com>
wrote:

> In pkg/build/registry/rest.go
> <https://github.com/openshift/origin/pull/4819#discussion_r40899072>:
>
> > +	"k8s.io/kubernetes/pkg/labels"
> > +
> > +	"github.com/openshift/origin/pkg/build/api"
> > +)
> > +
> > +// ErrUnknownBuildPhase is returned for WaitForRunningBuild if an unknown phase is returned.
> > +var ErrUnknownBuildPhase = fmt.Errorf("unknown build phase")
> > +
> > +// WaitForRunningBuild waits until the specified build is no longer New or Pending. Returns true if
> > +// the build became ter within timeout, false if it did not, and an error if any other error state occurred.
> > +// The last observed Build state is returned.
> > +func WaitForRunningBuild(watcher rest.Watcher, ctx kapi.Context, build *api.Build, timeout time.Duration) (*api.Build, bool, error) {
> > +	fieldSelector := fields.Set{"metadata.name": build.Name}.AsSelector()
> > +	w, err := watcher.Watch(ctx, labels.Everything(), fieldSelector, build.ResourceVersion)
> > +	if err != nil {
> > +		return nil, false, err
>
> We shouldn't return nil here. A build object containing a
> BuildPhaseUnknown status is also a cheap solution. I took a closer look
> and you don't use the latest build object except for its phase. So this
> func should return just the phase of the build.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4819/files#r40899072>.
>